core/src/main/scala/kafka/log/LogManager.scala

```diff
@@ -1176,6 +1176,42 @@ class LogManager(logDirs: Seq[File],
     }
   }
 
+  def recoverAbandonedFutureLogs(brokerId: Int, newTopicsImage: TopicsImage): Unit = {
+    val abandonedFutureLogs = findAbandonedFutureLogs(brokerId, newTopicsImage)
+    abandonedFutureLogs.foreach { case (futureLog, currentLog) =>
+      val tp = futureLog.topicPartition
+
+      futureLog.renameDir(UnifiedLog.logDirName(tp), shouldReinitialize = true)
+      futureLog.removeLogMetrics()
+      futureLogs.remove(tp)
+
+      currentLog.foreach { log =>
+        log.removeLogMetrics()
+        log.renameDir(UnifiedLog.logDeleteDirName(tp), shouldReinitialize = false)
+        addLogToBeDeleted(log)
+        info(s"Old log for partition ${tp} is renamed to ${log.dir.getAbsolutePath} and is scheduled for deletion")
+      }
+
+      currentLogs.put(tp, futureLog)
+      futureLog.newMetrics()
+
+      info(s"Successfully renamed abandoned future log for $tp")
```

Comment:
I can see we basically do the similar things as what `replaceCurrentWithFutureLog` did, why should we duplicate the codes here again? Could we not be able to call `replaceCurrentWithFutureLog` with additional parameters to indicate we want to skip some actions?

Like what we did now, we didn't close the `currentLog`, which should have potential resource leak. Using the `replaceCurrentWithFutureLog` could avoid this issue happening. WDYT?

Reply:
it seems to me that we need to call `closeHandlers` at least if we don't do refactor in this PR.

Reply:
That's a great suggestion! I'm going to try refactoring this to use `replaceCurrentWithFutureLog`

Reply:
I've refactored LogManager to use `replaceCurrentWithFutureLog` in [b87a21f](https://github.com/apache/kafka/pull/15136/commits/b87a21f3bd0eea6e4083a2d14b41053361f7b40a). PTAL